django/db/models/query.py

```diff
@@ -291,10 +291,11 @@ def __getitem__(self, k):
                 'QuerySet indices must be integers or slices, not %s.'
                 % type(k).__name__
             )
-        assert ((not isinstance(k, slice) and (k >= 0)) or
-                (isinstance(k, slice) and (k.start is None or k.start >= 0) and
-                 (k.stop is None or k.stop >= 0))), \
-            "Negative indexing is not supported."
+
+        if ((not isinstance(k, slice) and (k < 0)) or
+                (isinstance(k, slice) and ((k.start is not None and k.start < 0) or
+                 (k.stop is not None and k.stop < 0)))):
+            raise IndexError("Negative indexing is not supported.")
```

Comment:
Not the chief thrust of this PR but since we're touching these lines, we could consider single quoting these messages. Feel free to ignore unless someone else mentions it.